### DISCUSSION TITLE
How to add multiple subscriptions

### DISCUSSION THREAD

**Discussion by jimman100** [Azure/AzOps#793] [Q&A] [answered]
Hi, 

How do i add multiple subscriptions ? do i need to have multiple ARM template  file and parameters file each per subscriptions or can i have one arm template file and one parameters file with multiple values for subscriptions? 
for eg : do i need to have 
newsub1.json and newsub1.paramters.json and
newsub2.json and newsub2.paramters.json 

or can i have newsub.json and newsub.paramters.json, with the values for sub1 and sub2 specified in newsub.parameters.json ?

thanks
Jimmy

**Comment by Jefajers** [comment]
Hi if I understand your scenario correctly you are looking at using AzOps to create new subscriptions.
Perhaps inspired by something like [this](https://github.com/Azure/Enterprise-Scale/wiki/Create-Landingzones#create-a-new-landing-zone-subscriptions).

Yes you could have multiple subscriptions created from single `newsub.json` file. However, remember that based on file location their management group hierarchy target is selected. If you intend subscription a) and b) belong to different management groups you would need multiple files placed in different management group folders.

**Reply by jimman100** [reply]
Thanks ! Yes, all subscriptions will be under the same management group only( Will have one one file per management group). What I am looking for is how will be the parameter file look like if it is just one parameters file. I would like to keep one file and create the pull request against that one file when there is  request for a new subscription. Say today i get a request for one subscription, then i add the values in the file and tomorrow when i get a request for another subscription, i would like to append the values to the same values and AzOps should create the new subscription when the PR is merged. Will it be something like this in a single file ? I guess i am looking for the format to specify multiple value in the same file.

{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {
        "subscriptionAliasName": {
            "value": "subA" 
        },
        "billingAccountId": {
            "value": "" 
        },
        "targetManagementGroup": {
            "value": "" 
        }
    },
   "parameters": {
        "subscriptionAliasName": {
            "value": "subB" 
        },
        "billingAccountId": {
            "value": "" 
        },
        "targetManagementGroup": {
            "value": "" 
        }
    },
}
in the case above subA was created yesterday and today when i get a request, i need to add the paramter values in the same file so that subB can be created when the PR is merged. 
Thanks 
Jimmy

**Reply by Jefajers** [reply]
Got it, this scenario is dependent on standard ARM [functions.](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/copy-resources)

You could achieve what you are considering in several ways.

The “simplest” way would be to deploy only one subscription per PR. Then you could just replace the value of `subscriptionAliasName`.

However, this would not align with your first ask, “create multiple subscriptions per PR with one template and one parameter file.

As an example you could structure the parameter file like this:
```
{
	"$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
	"contentVersion": "1.0.0.0",
	"parameters": {
		"subscriptions": [{
				"subscriptionAliasName": "subname1",
				"billingAccountId": "345353454",
				"targetManagementGroup": "abc32"
			},
			{
				"subscriptionAliasName": "subname2",
				"billingAccountId": "345353454",
				"targetManagementGroup": "abc32"
			}
		]
	}
}
```
And have a look at the ARM copy function and update the deployment template to make use of it.

```
{
	"$schema": "https://schema.management.azure.com/schemas/2019-08-01/managementGroupDeploymentTemplate.json#",
	"contentVersion": "1.0.0.0",
	"parameters": {
		"subscriptions": {
			"type": "object"
		}
	},
	"resources": [{
		"copy": {
			"name": "subcopy",
			"count": "[length(parameters('subscriptions'))]"
		},
		"scope": "/", // routing the request to tenant root
		"name": "[parameters('subscriptions').subscriptionAliasName[copyIndex()]]",
		"type": "Microsoft.Subscription/aliases",
		"apiVersion": "2020-09-01",
		"properties": {
			"workLoad": "Production",
			"displayName": "[parameters('subscriptions').subscriptionAliasName[copyIndex()]]",
			"billingScope": "[parameters('subscriptions').billingAccountId[copyIndex()]]",
			"managementGroupId": "[tenantResourceId('Microsoft.Management/managementGroups/', parameters('subscriptions').targetManagementGroup[copyIndex()])]"
		}
	}],
	"outputs": {}
}
```

**Comment by jimman100** [comment]
Thanks for the answer. Solution with the copy function looks good. I have another question though. My intention is not to create multiple subscription in one go. With one PR, I expect to create only a single subscription.  I want to maintain the subscription values in the parameter file so that i can see the changes to the file in the history. With the copy function, say if the first two subscriptions are created today, can i use the same file and and the values for the third subscription in the same file so that when the pr is merged, it creates only the third subscription as the first two subscriptions were already created previously. 
In other words, can i keep appending the parameter with each pull request in the paramters.json file (instead of creating a new file each time ?)

With your first suggestion to replace the value of subscriptionAliasName with each PR, I need to keep previous subscriptionAliasName also in the same file so that i have the history of the changes. Is there a way to keep the previous subscriptionAliasName also in the same file and to add the new subscriptionAliasName so that PR will not fail as one of the subscriptionAliasName already exists? 

Thanks again

**Comment by Jefajers** [comment, marked as answer]
> In other words, can i keep appending the parameter with each pull request in the paramters.json file (instead of creating a new file each time ?)

Yes, you can keep appending parameters with each pull request. [ARM templates are idempotent](https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/overview#why-choose-arm-templates), which means you can deploy the same template many times and get the same resource types in the same state.

**Reply by jimman100** [reply]
Thanks, my apologies for the delay in getting back to you. I am getting an error when using the copy function. instead can you please let me know if this is possible. i will have one newSubscription.json file which is already in the repo, when i want to create a new subscription, i will create a new parameter file and merge it. please note, i will only check-in the parameter file, can the workflow, create the subscription based on just the parameter file or do i need to check-in the template file also along with the parameter file each time (2 files per PR) to create the subscription ? 

say newSubscription.json is already in the repo, i will then create newSubscription.1.paramters.json and create a PR with justnewSubscription.1.paramters.json and when i merge the PR, can it create the subscription. I tried this and the validate workflow had failed.

Does it need to be always 2 files  per PR for the subscription to be created ?
Thanks
Jimmy

**Reply by jimman100** [reply]
sorry, missed to add the error message with the validate workflow when just the paramters file is merged as part of the PR


WARNING: [06:05:49][Invoke-AzOpsPush] effectiveResourceType:  AzOpsMainTemplate does NOT supports resource type  in /home/runner/.local/share/powershell/Modules/AzOps/2.1.2/data/template/template.json. Deployment will be ignored
[06:05:49][Invoke-AzOpsPush] Expecting deploymentList object, it was not found. Error exiting runtime.
Exception: ScriptHalted

Error: Process completed with exit code 1.
